akd_local_auditor/src/auditor/mod.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under both the MIT license found in the
+// LICENSE-MIT file in the root directory of this source tree and the Apache
+// License, Version 2.0 found in the LICENSE-APACHE file in the root directory
+// of this source tree.
+
+//! This module holds the auditor operations based on binary-encoded AuditProof blobs
+
+use anyhow::{anyhow, bail, Result};
+use clap::{Parser, Subcommand};
+use log::{debug, error, info, warn};
+use std::marker::{Send, Sync};
+
+use rustyline::error::ReadlineError;
+use rustyline::Editor;
+
+const HISTORY_FILE: &str = ".akd_local_auditor_history";
+
+fn format_qr_record<H>(p_hash: H::Digest, c_hash: H::Digest, epoch: u64) -> Vec<u8>
+where
+    H: winter_crypto::Hasher,
+{
+    let p_bytes = akd::serialization::from_digest::<H>(p_hash);
+    let c_bytes = akd::serialization::from_digest::<H>(c_hash);
+    let epoch_bytes = epoch.to_ne_bytes();
+    let header = "WA_AKD_VERIFY".as_bytes();
+
+    let mut result = vec![];
+    result.extend(header);
+    result.extend(epoch_bytes);
+    result.extend(p_bytes);
+    result.extend(c_bytes);
+    result
+}
+
+pub async fn audit_epoch<H>(blob: akd::proto::AuditBlob, qr: bool) -> Result<()>
+where
+    H: winter_crypto::Hasher + Sync + Send,
+{
+    let (p_hash, c_hash, epoch, proof) = blob.decode().map_err(|err| anyhow!("{}", err))?;
+
+    // DO THE FREAKING VERIFICATION ALREADY
+    if let Err(akd_error) = akd::auditor::audit_verify(
+        vec![p_hash, c_hash],
+        akd::proof_structs::AppendOnlyProof::<crate::Hasher> {
+            proofs: vec![proof],
+            epochs: vec![epoch],
+        },
+    )
+    .await
+    {
+        warn!(
+            "Audit proof for epoch {} failed to verify with error {}",
+            epoch, akd_error
+        );
+    } else {
+        info!("Audit proof for epoch {} has verified!", epoch);
+        if qr {
+            info!("Generating scan-able QR code for the verification on device");
+            let qr_code_data = format_qr_record::<crate::Hasher>(p_hash, c_hash, epoch);
+            if let Err(error) = qr2term::print_qr(&qr_code_data) {
+                error!("Error generating QR code {}", error);
+                bail!("Error generating QR code {}", error);
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Storage options for retrieving audit proofs
+#[derive(Subcommand, Clone, Debug)]
+pub enum AuditCommand {
+    /// Audit a specific epoch
+    Audit {
+        /// The epoch to audit
+        #[clap(long)]
+        epoch: u64,
+        /// Show a QR code of the audit
+        #[clap(long)]
+        qr: bool,
```

Comment:
This is really cool! Only one tiny suggestion: should we make this indicative of the flag that it's a flag, e.g., `disp_qr` or something? Dont' want to really make it verbose, but we can always specify a short-flag for it if that's a concern.

Reply:
I'm good with `qr`!